src/matrix/room/timeline/Timeline.js

```diff
@@ -236,7 +251,128 @@ export class Timeline {
     /** @package */
     addEntries(newEntries) {
         this._addLocalRelationsToNewRemoteEntries(newEntries);
+        this._updateEntriesNotInTimeline(newEntries);
+        this._moveEntryToRemoteEntries(newEntries);
         this._remoteEntries.setManySorted(newEntries);
+        this._loadContextEntriesWhereNeeded(newEntries);
+    }
+
+    /**
+     * Update entries based on newly received events.
+     * eg: a newly received redaction event may mark an existing event in contextEntriesNotInTimeline as being redacted
+     * This is only for the events that are not in the timeline but had to fetched from elsewhere to render reply previews.
+     */
+    _updateEntriesNotInTimeline(entries) {
+        for (const entry of entries) {
+            const relatedEntry = this._contextEntriesNotInTimeline.get(entry.relatedEventId);
```

Comment:
This doesn't seem to take into account that a context entry that came from storage (but not in the timeline) will still be updated by the syncwriter, and the entry that is passed in will actually be the relation target.

So let's say we have a message E. Then there is a reply F to message E. E wasn't in the timeline, so is put into _contextEntriesNotInTimeline. Now a redaction G is sent to redact message E. When G is received by our client through sync, E is known in storage. So the sync writer, delegating to the relation writer, will fetch and update E from storage, and thus an already redacted version of E will be passed to replaceEntries, while G will be passed to addEntries.

I'm not sure it makes sense to call the same method from both addEntries and replaceEntries for this purpose. I think we should do what you already do for events that are not in storage (e.g. fetched from the homeserver) in addEntries, but we should also look for entry.id in _contextEntriesNotInTimeline for the above case in replaceEntries.

Does that make sense?